cibuildwheel/linux.py

```diff
@@ -158,11 +62,132 @@ def build(project_dir, output_dir, test_command, test_requires, test_extras, bef
+                            container_name + ':/project'], check=True)
+
+            for config in platform_configs:
+                if dependency_constraints:
+                    constraints_file = dependency_constraints.get_for_python_version(config.version)
+                    subprocess.run(['docker', 'cp',
+                                    os.path.abspath(constraints_file),
+                                    container_name + ':/constraints.txt'], check=True)
+
+                subprocess.run(
+                    ['docker', 'start', '-i', '-a', container_name],
+                    universal_newlines=True,
+                    check=True,
+                    input='''
+                        set -o errexit
+                        set -o xtrace
+                        mkdir -p /output
+                        cd /project
+
+                        PYBIN="{config_python_bin}"
+                        export PATH="$PYBIN:$PATH"
+
+                        {environment_exports}
+
+                        # check the active python and pip are in PYBIN
+                        # if `test` returns false, the script will exit due to errexit
+                        test "$(which pip)" = "$PYBIN/pip"
+                        test "$(which python)" = "$PYBIN/python"
+
+                        if [ ! -z {before_build} ]; then
+                            sh -c {before_build}
+                        fi
+
+                        # Build the wheel
+                        rm -rf /tmp/built_wheel
+                        mkdir /tmp/built_wheel
+                        pip wheel . -w /tmp/built_wheel --no-deps {build_verbosity_flag}
+                        built_wheel=(/tmp/built_wheel/*.whl)
+
+                        # repair the wheel
+                        rm -rf /tmp/repaired_wheels
+                        mkdir /tmp/repaired_wheels
+                        # NOTE: 'built_wheel' here is a bash array of glob matches; "$built_wheel" returns
+                        # the first element
+                        if [[ "$built_wheel" == *none-any.whl ]] || [ -z {repair_command} ]; then
+                            # pure Python wheel or empty repair command
+                            mv "$built_wheel" /tmp/repaired_wheels
+                        else
+                            sh -c {repair_command} repair_command "$built_wheel"
+                        fi
+                        repaired_wheels=(/tmp/repaired_wheels/*.whl)
+
+                        if [ ! -z {test_command} ]; then
+                            # Set up a virtual environment to install and test from, to make sure
+                            # there are no dependencies that were pulled in at build time.
+                            pip install {dependency_install_flags} virtualenv
+                            venv_dir=`mktemp -d`/venv
+                            python -m virtualenv --no-download "$venv_dir"
```

Comment:
Hmm. Test if this venv is created properly may be harder than in macos or windows.

Reply:
--no-download flag should force to have same version of packages in main python and in virtualenv. If we would like to test if this work this way then there is need to determine package version. But in manylinux cibuildwheel do not manage package version so test should not use version saved in requirements but determine it in runtime (it should be different test that for windows and macos)

Reply:
Ah, yes, I see, I think. Good point!

Reply:
`--no-download` causes virtualenv to install versions of setuptools, pip that were vendored into that version of virtualenv, I think.

![image](https://user-images.githubusercontent.com/1244307/75607550-29504a00-5af0-11ea-82b2-c905e67b3c77.png)

I'm not sure how to test that, since the version installed is dependent on that version of virtualenv. I suppose we could hard-code it into a test, but that would change after running `bin/update_constraints`.

Since it's quite hard to test, my inclination is to just trust that virtualenv's behaviour remains deterministic when used with `--no-download`. After all, the versions in the test virtualenv don't affect the wheel that's built, anyway.